inference-engine/src/multi_device/multi_device_exec_network.cpp

```diff
@@ -159,109 +159,181 @@ MultiDeviceExecutableNetwork::MultiDeviceExecutableNetwork(const std::string&
+    }
+    catch (const std::exception& e) {
+        return;
+    }
+
+    // if selec device is CPU, do not need to load CPU again, context[CPU] must have loaded CPU
+    curDevIsCPU = (context.deviceInfo.deviceName.find("CPU") != std::string::npos);
+    if (curDevIsCPU) {
+        return;
+    }
+
+    // try to load this candidate device
+    TryToLoadNetWork(context, modelPath, network);
 }
 
 void MultiDeviceExecutableNetwork::WaitFirstNetworkReady() {
-    if (_alreadyActualNetwork) {
-        return;
+    if (_firstLoadFuture.valid()) {
+        // wait for the first loading finished
+        _firstLoadFuture.wait();
     }
-    if (_cpuFuture.valid() && _acceleratorFuture.valid()) {
-        try {
-            _networkFirstReady = _cpuFuture.get();
-        } catch (const std::exception& e) {
-            printf("Warning: load network to CPU failed: %s\n", e.what());
-            _networkActualNeeded = _acceleratorFuture.get();
+
+    // check if there is any device that have loaded network successfully
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled && _loadContext[i].isAlready) {
+            return;
+        }
+    }
+
+    // the first loading is failed, wait for another loading
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled) {
+            _loadContext[i].future.wait();
         }
-    } else if (_acceleratorFuture.valid()) {  // only accelerator is valid, like AUTO:GPU
-        _networkActualNeeded = _acceleratorFuture.get();
-    } else if (_cpuFuture.valid()) {  // only CPU is valid, like AUTO:CPU
-        _networkActualNeeded = _cpuFuture.get();
-    } else {
-        IE_THROW() << "No device task available";
     }
 
-    // if there is only one device or loading CPU device is failed,
-    // the ActualNetwork is already ok now.
-    if (!_acceleratorFuture.valid()) {
-        _alreadyActualNetwork = true;
+    // check if another loading is successful
+    for (int i = CONTEXTNUM - 1; i >= 0; i--) {
+        if (_loadContext[i].isEnabled && _loadContext[i].isAlready) {
+            return;
+        }
     }
+
+    // To Do need to print failed error mesage
+    IE_THROW() << "[AUTO] load all devices failed";
 }
 
 void MultiDeviceExecutableNetwork::WaitActualNetworkReady() const {
+    OV_ITT_SCOPED_TASK(itt::domains::MULTIPlugin, "MultiDeviceExecutableNetwork::WaitActualNetworkReady");
     // Maybe different API will call this function, so add call once here
     // for every MultiDeviceExecutableNetwork instance
-    OV_ITT_SCOPED_TASK(itt::domains::MULTIPlugin, "MultiDeviceExecutableNetwork::WaitActualNetworkReady");
-    std::call_once(_oc, [&] () {
-            if (_acceleratorFuture.valid()) {
-                _networkActualNeeded = _acceleratorFuture.get();
-            }
+    std::call_once(_oc, [this] () {
+               if (_loadContext[ACTUALDEVICE].future.valid()) {
+                   _loadContext[ACTUALDEVICE].future.get();
```

Comment:
if an exception is not handle, it will cause the application exit.
if promise set_exception,  it will cause future.get() to re-thrown exception
in this code, we handled the failed load exception, but not set exception, so it will not re-thrown

Reply:
in the destructor it is may be good not to re-throw anything , as if there is another exception being propagated (e.g. the actual reason of the app doing the exit) the terminate will be called.